cmd/add_test.go

```diff
@@ -0,0 +1,122 @@
+`
+
+//nolint:lll // no need to check for line length.
+var indexAddHelp = `Add an index to the local falcoctl configuration. Indexes are used to perform search operations for artifacts
+
+Usage:
+  falcoctl index add [NAME] [URL] [BACKEND] [flags]
+
+Flags:
+  -h, --help   help for add
+
+Global Flags:
+      --config string     config file to be used for falcoctl (default "/etc/falcoctl/falcoctl.yaml")
+      --disable-styling   Disable output styling such as spinners, progress bars and colors. Styling is automatically disabled if not attacched to a tty (default false)
+  -v, --verbose           Enable verbose logs (default false)
+`
+
+var addAssertFailedBehavior = func(usage, specificError string) {
+	It("check that fails and the usage is not printed", func() {
+		Expect(err).To(HaveOccurred())
+		Expect(output).ShouldNot(gbytes.Say(regexp.QuoteMeta(usage)))
+		Expect(output).Should(gbytes.Say(regexp.QuoteMeta(specificError)))
+	})
+}
+
+var indexAddTests = Describe("add", func() {
+
+	var (
+		indexCmd  = "index"
+		addCmd    = "add"
+		indexName = "testName"
+	)
+
+	// Each test gets its own root command and runs it.
+	// The err variable is asserted by each test.
+	JustBeforeEach(func() {
+		rootCmd = cmd.New(ctx, opt)
+		err = executeRoot(args)
+	})
+
+	JustAfterEach(func() {
+		Expect(output.Clear()).ShouldNot(HaveOccurred())
+	})
+
+	Context("help message", func() {
+		BeforeEach(func() {
+			args = []string{indexCmd, addCmd, "--help"}
+		})
+
+		It("should match the saved one", func() {
+
+			Expect(output).Should(gbytes.Say(regexp.QuoteMeta(indexAddHelp)))
+		})
+	})
+
+	// Here we are testing all the failure cases using both the rulesfile and plugin artifact types.
+	// The common logic for the artifacts is tested once using a rulesfile artifact, no need to repeat
+	// the same test using a plugin artifact.
+	Context("failure", func() {
+		When("without URL", func() {
+			BeforeEach(func() {
+				args = []string{indexCmd, addCmd, "--config", configFile, indexName}
+			})
+			addAssertFailedBehavior(indexAddUsage, "ERRO: accepts between 2 and 3 arg(s), received 1")
+		})
+
+		When("with invalid URL", func() {
+			BeforeEach(func() {
+				args = []string{indexCmd, addCmd, "--config", configFile, indexName, "NOTAPROTOCAL://something"}
+			})
+			addAssertFailedBehavior(indexAddUsage, "ERRO: unable to add index: unable to fetch index \"testName\""+
+				" with URL \"NOTAPROTOCAL://something\": unable to fetch index: cannot fetch index: Get "+
+				"\"notaprotocal://something\": unsupported protocol scheme \"notaprotocal\"")
+		})
+
+		When("with invalid backend", func() {
+			BeforeEach(func() {
+				args = []string{indexCmd, addCmd, "--config", configFile, indexName, "http://noindex", "notabackend"}
+			})
+			addAssertFailedBehavior(indexAddUsage, "ERRO: unable to add index: unable to fetch index \"testName\" "+
+				"with URL \"http://noindex\": unsupported index backend type: notabackend")
+		})
```

Comment:
I only added failure tests for the new behavior atm, because i noticed that the current implementation of
`falcoctl index add` behaves strangely or could be considered buggy i.e.,

If you call `index add` for and `index name` for which there already exists in the `cache` the CMD will **always** return `Index %q successfully added` and do no modification. For example  consider the following sequence

```
$ cat example.yaml                                                                                      
cat: example.yaml: No such file or directory
$ ./falcoctl index add --config example.yaml example https://falcosecurity.github.io/falcoctl/index.yaml
 INFO  Adding index
 INFO  Index "example" successfully added
$ cat example.yaml                                                                                      
indexes:
    - name: falcosecurity
      url: https://falcosecurity.github.io/falcoctl/index.yaml
      backend: ""
    - name: example
      url: https://falcosecurity.github.io/falcoctl/index.yaml
      backend: ""
$ ./falcoctl index add --config example.yaml example NOPE://NOPE                                        
 INFO  Adding index
 INFO  Index "example" successfully added
$ cat example.yaml                                                                                      
indexes:
    - name: falcosecurity
      url: https://falcosecurity.github.io/falcoctl/index.yaml
      backend: ""
    - name: example
      url: https://falcosecurity.github.io/falcoctl/index.yaml
      backend: ""
 ```

 One would expect the add command to error that the index already exists or that the index is invalid.

Reply:
Hi @max-frank, the initial idea was to have the `add command `return an error only in cases when there is not possible to add an index. I agree with you that the message should be different if the index already exists, but it should not error. 

Anyway, feel free to modify the current implementation and extend the tests.

Reply:
Understood if its modified behavior I can modify the output in a follow up PR.

I'll try to add more tests relevant to this PR I initial did not because of confusion regarding the above interaction and that I a wall in making the cache and config clean up work correctly in between test runs.

Reply:
@alacuku For the tests i have run into a slight problem here. For the add tests to be stable and isolated I need to make sure to clean all index config and cache files after each run, but the location of those files is right now not configurable (see [code here](https://github.com/falcosecurity/falcoctl/blob/main/internal/config/config.go#L149-L152)). 

Now since I cant use an isolated tmp config dir it would mean that `index add` tests would have side effects on developer machines, i.e., Indices from tests are added to their real local configuration and if we were to do a clean up after a test case we would delete all their index config.

Is it ok if i skip the positive tests here for now until this issue is addressed?